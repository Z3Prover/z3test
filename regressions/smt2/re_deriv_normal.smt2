(declare-const c1 (_ BitVec 8))
(declare-const x String)

(echo "====== Example derivative normal forms ======")

(echo "==== ranges ====")

(echo "== empty range ==")
(simplify (re.derivative c1 (re.range "b" "a")))
(echo "== single character range ==")
(simplify (re.derivative c1 (re.range "a" "a")))
(echo "== multiple character range ==")
(simplify (re.derivative c1 (re.range "a" "b")))
(echo "== range star ==")
(simplify (re.derivative c1 (re.* (re.range "a" "b"))))
(echo "== union of nested ranges ==")
(simplify (re.derivative c1 (re.union (re.range "a" "g") (re.range "c" "e"))))
(echo "== union of disjoint ranges ==")
(simplify (re.derivative c1 (re.union (re.range "a" "c") (re.range "e" "g"))))
(echo "== union of intersecting ranges ==")
(simplify (re.derivative c1 (re.union (re.range "a" "e") (re.range "c" "g"))))
(echo "== intersection of nested ranges ==")
(simplify (re.derivative c1 (re.inter (re.range "a" "g") (re.range "c" "e"))))
(echo "== intersection of disjoint ranges ==")
(simplify (re.derivative c1 (re.inter (re.range "a" "c") (re.range "e" "g"))))
(echo "== intersection of intersecting ranges ==")
(simplify (re.derivative c1 (re.inter (re.range "a" "e") (re.range "c" "g"))))

(echo "==== characters/strings ====")

(echo "== single string ==")
(simplify (re.derivative c1 (str.to_re "cat")))
(echo "== union of characters ==")
(simplify (re.derivative c1 (re.union (str.to_re "a") (re.union (str.to_re "b") (str.to_re "c")))))
(echo "== union of strings starting with the same character ==")
(simplify (re.derivative c1 (re.union (str.to_re "at") (re.union (str.to_re "are") (str.to_re "am")))))
(echo "== union of strings starting with the different characters ==")
(simplify (re.derivative c1 (re.union (str.to_re "to") (re.union (str.to_re "if") (str.to_re "a")))))
(echo "== large union of various strings ==")
(simplify (re.derivative c1 (re.union (str.to_re "four") (re.union (str.to_re "score") (re.union (str.to_re "and") (re.union (str.to_re "seven") (re.union (str.to_re "years") (str.to_re "ago"))))))))

(echo "==== complex examples ====")

(echo "== 1 ==")
(simplify (re.derivative c1 (re.inter (re.union (re.* (re.range "a" "b")) (re.* (re.range "c" "d"))) (re.* (re.range "b" "c")))))
(echo "== 2 ==")
(simplify (re.derivative c1 (re.inter (re.++ re.all (re.range "0" "0")) (re.* (re.range "1" "1")))))

(echo "==== nullable infinite loop checks ====")

(echo "== loop 1 (original) ==")
(simplify (str.in_re "" (re.inter (re.comp (re.derivative c1 (re.derivative (seq.nth_i x 0) (str.to_re x)))) (re.comp (re.derivative c1 (str.to_re x))))))
(echo "== loop 2 (minimized) ==")
(simplify (str.in_re "" (re.inter (re.derivative c1 (re.derivative c1 (str.to_re x))) (re.derivative c1 (str.to_re x)))))
