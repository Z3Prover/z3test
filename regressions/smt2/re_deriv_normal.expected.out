====== Example derivative normal forms ======
==== ranges ====
== empty range ==
re.none
== single character range ==
(let ((a!1 (ite (and (= ((_ extract 7 7) c1) #b0)
                     (bvule ((_ extract 6 0) c1) #b1100000))
                re.none
                (str.to_re ""))))
  (ite (and (= ((_ extract 7 7) c1) #b0) (bvule ((_ extract 6 0) c1) #b1100001))
       a!1
       re.none))
== multiple character range ==
(let ((a!1 (ite (and (= ((_ extract 7 7) c1) #b0)
                     (bvule ((_ extract 6 0) c1) #b1100000))
                re.none
                (str.to_re ""))))
  (ite (and (= ((_ extract 7 7) c1) #b0) (bvule ((_ extract 6 0) c1) #b1100010))
       a!1
       re.none))
== range star ==
(let ((a!1 (ite (and (= ((_ extract 7 7) c1) #b0)
                     (bvule ((_ extract 6 0) c1) #b1100000))
                re.none
                (re.* (re.range "a" "b")))))
  (ite (and (= ((_ extract 7 7) c1) #b0) (bvule ((_ extract 6 0) c1) #b1100010))
       a!1
       re.none))
== union of nested ranges ==
(let ((a!1 (ite (and (= ((_ extract 7 7) c1) #b0)
                     (bvule ((_ extract 6 0) c1) #b1100000))
                re.none
                (str.to_re ""))))
(let ((a!2 (ite (and (= ((_ extract 7 7) c1) #b0)
                     (bvule ((_ extract 6 0) c1) #b1100010))
                a!1
                (str.to_re ""))))
(let ((a!3 (ite (and (= ((_ extract 7 7) c1) #b0)
                     (bvule ((_ extract 6 0) c1) #b1100101))
                a!2
                (str.to_re ""))))
  (ite (and (= ((_ extract 7 7) c1) #b0) (bvule ((_ extract 6 0) c1) #b1100111))
       a!3
       re.none))))
== union of disjoint ranges ==
(let ((a!1 (ite (and (= ((_ extract 7 7) c1) #b0)
                     (bvule ((_ extract 6 0) c1) #b1100000))
                re.none
                (str.to_re ""))))
(let ((a!2 (ite (and (= ((_ extract 7 7) c1) #b0)
                     (bvule ((_ extract 6 0) c1) #b1100011))
                a!1
                re.none)))
(let ((a!3 (ite (and (= ((_ extract 7 7) c1) #b0)
                     (bvule ((_ extract 6 0) c1) #b1100100))
                a!2
                (str.to_re ""))))
  (ite (and (= ((_ extract 7 7) c1) #b0) (bvule ((_ extract 6 0) c1) #b1100111))
       a!3
       re.none))))
== union of intersecting ranges ==
(let ((a!1 (ite (and (= ((_ extract 7 7) c1) #b0)
                     (bvule ((_ extract 6 0) c1) #b1100000))
                re.none
                (str.to_re ""))))
(let ((a!2 (ite (and (= ((_ extract 7 7) c1) #b0)
                     (bvule ((_ extract 6 0) c1) #b1100010))
                a!1
                (str.to_re ""))))
(let ((a!3 (ite (and (= ((_ extract 7 7) c1) #b0)
                     (bvule ((_ extract 6 0) c1) #b1100101))
                a!2
                (str.to_re ""))))
  (ite (and (= ((_ extract 7 7) c1) #b0) (bvule ((_ extract 6 0) c1) #b1100111))
       a!3
       re.none))))
== intersection of nested ranges ==
(let ((a!1 (ite (and (= ((_ extract 7 7) c1) #b0)
                     (bvule ((_ extract 6 0) c1) #b1100010))
                re.none
                (str.to_re ""))))
(let ((a!2 (ite (and (= ((_ extract 7 7) c1) #b0)
                     (bvule ((_ extract 6 0) c1) #b1100101))
                a!1
                re.none)))
  (ite (and (= ((_ extract 7 7) c1) #b0) (bvule ((_ extract 6 0) c1) #b1100111))
       a!2
       re.none)))
== intersection of disjoint ranges ==
re.none
== intersection of intersecting ranges ==
(let ((a!1 (ite (and (= ((_ extract 7 7) c1) #b0)
                     (bvule ((_ extract 6 0) c1) #b1100010))
                re.none
                (str.to_re ""))))
(let ((a!2 (ite (and (= ((_ extract 7 7) c1) #b0)
                     (bvule ((_ extract 6 0) c1) #b1100101))
                a!1
                re.none)))
  (ite (and (= ((_ extract 7 7) c1) #b0) (bvule ((_ extract 6 0) c1) #b1100111))
       a!2
       re.none)))
==== characters/strings ====
== single string ==
(let ((a!1 (ite (and (= ((_ extract 7 7) c1) #b0)
                     (bvule ((_ extract 6 0) c1) #b1100010))
                re.none
                (str.to_re "at"))))
  (ite (and (= ((_ extract 7 7) c1) #b0) (bvule ((_ extract 6 0) c1) #b1100011))
       a!1
       re.none))
== union of characters ==
(let ((a!1 (ite (and (= ((_ extract 7 7) c1) #b0)
                     (bvule ((_ extract 6 0) c1) #b1100000))
                re.none
                (str.to_re ""))))
(let ((a!2 (ite (and (= ((_ extract 7 7) c1) #b0)
                     (bvule ((_ extract 6 0) c1) #b1100001))
                a!1
                (str.to_re ""))))
(let ((a!3 (ite (and (= ((_ extract 7 7) c1) #b0)
                     (bvule ((_ extract 6 0) c1) #b1100010))
                a!2
                (str.to_re ""))))
  (ite (and (= ((_ extract 7 7) c1) #b0) (bvule ((_ extract 6 0) c1) #b1100011))
       a!3
       re.none))))
== union of strings starting with the same character ==
(let ((a!1 (ite (and (= ((_ extract 7 7) c1) #b0)
                     (bvule ((_ extract 6 0) c1) #b1100000))
                re.none
                (re.union (str.to_re "t") (str.to_re "re") (str.to_re "m")))))
  (ite (and (= ((_ extract 7 7) c1) #b0) (bvule ((_ extract 6 0) c1) #b1100001))
       a!1
       re.none))
== union of strings starting with the different characters ==
(let ((a!1 (ite (and (= ((_ extract 7 7) c1) #b0)
                     (bvule ((_ extract 6 0) c1) #b1100000))
                re.none
                (str.to_re ""))))
(let ((a!2 (ite (and (= ((_ extract 7 7) c1) #b0)
                     (bvule ((_ extract 6 0) c1) #b1100001))
                a!1
                re.none)))
(let ((a!3 (ite (and (= ((_ extract 7 7) c1) #b0)
                     (bvule ((_ extract 6 0) c1) #b1101000))
                a!2
                (str.to_re "f"))))
(let ((a!4 (ite (and (= ((_ extract 7 7) c1) #b0)
                     (bvule ((_ extract 6 0) c1) #b1101001))
                a!3
                re.none)))
(let ((a!5 (ite (and (= ((_ extract 7 7) c1) #b0)
                     (bvule ((_ extract 6 0) c1) #b1110011))
                a!4
                (str.to_re "o"))))
  (ite (and (= ((_ extract 7 7) c1) #b0) (bvule ((_ extract 6 0) c1) #b1110100))
       a!5
       re.none))))))
== large union of various strings ==
(let ((a!1 (ite (and (= ((_ extract 7 7) c1) #b0)
                     (bvule ((_ extract 6 0) c1) #b1100000))
                re.none
                (re.union (str.to_re "go") (str.to_re "nd")))))
(let ((a!2 (ite (and (= ((_ extract 7 7) c1) #b0)
                     (bvule ((_ extract 6 0) c1) #b1100001))
                a!1
                re.none)))
(let ((a!3 (ite (and (= ((_ extract 7 7) c1) #b0)
                     (bvule ((_ extract 6 0) c1) #b1100101))
                a!2
                (str.to_re "our"))))
(let ((a!4 (ite (and (= ((_ extract 7 7) c1) #b0)
                     (bvule ((_ extract 6 0) c1) #b1100110))
                a!3
                re.none)))
(let ((a!5 (ite (and (= ((_ extract 7 7) c1) #b0)
                     (bvule ((_ extract 6 0) c1) #b1110010))
                a!4
                (re.union (str.to_re "core") (str.to_re "even")))))
(let ((a!6 (ite (and (= ((_ extract 7 7) c1) #b0)
                     (bvule ((_ extract 6 0) c1) #b1110011))
                a!5
                re.none)))
(let ((a!7 (ite (and (= ((_ extract 7 7) c1) #b0)
                     (bvule ((_ extract 6 0) c1) #b1111000))
                a!6
                (str.to_re "ears"))))
  (ite (and (= ((_ extract 7 7) c1) #b0) (bvule ((_ extract 6 0) c1) #b1111001))
       a!7
       re.none))))))))
==== complex examples ====
== 1 ==
(let ((a!1 (ite (and (= ((_ extract 7 7) c1) #b0)
                     (bvule ((_ extract 6 0) c1) #b1100001))
                re.none
                (re.inter (re.* (re.range "a" "b")) (re.* (re.range "b" "c"))))))
(let ((a!2 (ite (and (= ((_ extract 7 7) c1) #b0)
                     (bvule ((_ extract 6 0) c1) #b1100010))
                a!1
                (re.inter (re.* (re.range "c" "d")) (re.* (re.range "b" "c"))))))
(let ((a!3 (ite (and (= ((_ extract 7 7) c1) #b0)
                     (bvule ((_ extract 6 0) c1) #b1100011))
                a!2
                re.none)))
  (ite (and (= ((_ extract 7 7) c1) #b0) (bvule ((_ extract 6 0) c1) #b1100100))
       a!3
       re.none))))
== 2 ==
(let ((a!1 (ite (and (= ((_ extract 7 6) c1) #b00)
                     (bvule ((_ extract 5 0) c1) #b110000))
                re.none
                (re.inter (re.* (re.range "1" "1"))
                          (re.++ re.all (re.range "0" "0"))))))
  (ite (and (= ((_ extract 7 6) c1) #b00) (bvule ((_ extract 5 0) c1) #b110001))
       a!1
       re.none))
==== nullable infinite loop checks ====
== loop 1 (original) ==
(let ((a!1 (not (str.in_re "" (re.derivative c1 (str.to_re x)))))
      (a!2 (str.in_re ""
                      (re.derivative c1
                                     (re.derivative (seq.nth_i x 0)
                                                    (str.to_re x))))))
  (and a!1 (not a!2)))
== loop 2 (minimized) ==
(let ((a!1 (str.in_re "" (re.derivative c1 (re.derivative c1 (str.to_re x))))))
  (and a!1 (str.in_re "" (re.derivative c1 (str.to_re x)))))
